docs/design/variadics.md

```diff
@@ -0,0 +1,929 @@
+code:
+
+```carbon
+fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {
+  let z: auto = (0 as f32, ... each x, ... each y);
+}
+```
+
+We proceed by finding the type of each segment. The type of `0 as f32` is `f32`,
+by the usual non-variadic typing rules. The type of `... each x` is
+`... Optional(each T)`, because `Optional(each T)` is the declared type of
+`each x`, and the type of a pack expansion is a pack expansion of the type of
+its body.
+
+The type of `... each y` is more complicated. Conceptually, it consists of some
+number of repetitions of `i32`. We don't know exactly how many repetitions,
+because it's implicitly specified by the caller: it's the arity of the second
+argument tuple. Effectively, that arity acts as a hidden deduced parameter of
+`F`.
+
+So to represent this type, we need two new pseudo-syntaxes:
+
+-   `‖each X‖` refers to the deduced arity of the pack expansion that contains
+    the declaration of `each X`.
+-   `⟪E; N⟫` evaluates to `N` repetitions of `E`. This is called a _arity
+    coercion_, because it coerces the expression `E` to have arity `N`. `E` must
+    not contain any pack expansions, each-names, or pack literals (see below).
+
+Combining the two, the type of `... each y` is `... ⟪i32; ‖each y‖⟫`. Thus, the
+type of `z` is `(f32, ... Optional(each T), ... ⟪i32; ‖each y‖⟫)`.
+
+Now, consider a modified version of that example:
+
+```carbon
+fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {
+  let (.. each z: auto) = (0 as f32, ... each x, ... each y);
+}
+```
+
+`each z` is a pack, but it has the same elements as the tuple `z` in our earlier
+example, so we represent its type in the same way, as a sequence of segments:
+`⟬f32, Optional(each T), ⟪i32; ‖each y‖⟫⟭`. The `⟬⟭` delimiters make this a
+_pack literal_ rather than a tuple literal. Notice one subtle difference: the
+segments of a pack literal do not contain `...`. In effect, every segment of a
+pack literal acts as a separate loop body. As with the tuple literal syntax, the
+pack literal pseudo-syntax can also be used in patterns.
+
+The _shape_ of a pack literal is a tuple of the arities of its segments, so the
+shape of `⟬f32, Optional(each T), ⟪i32; ‖each y‖⟫⟭` is
+`(1, ‖each T‖, ‖each y‖)`. Other expressions and patterns also have shapes. In
+particular, the shape of an arity coercion `⟪E; A⟫` is `(A)`, the shape of
+`each X` is `‖each X‖`, and the shape of an expression that does not contain
+pack literals, shape coercions, or each-names is 1. The arity of an expression
+is the sum of the elements of its shape. See the
+[appendix](#typing-and-shaping-rules) for the full rules for determining the
+shape of an expression.
+
+A pack literal can be _expanded_, which moves its parent AST node inside the
+pack literal, so long as the parent node is not `...`. For example,
+`... Optional(⟬each X, Y⟭)` is equivalent to
+`... ⟬Optional(each X), Optional(Y)⟭`. Similarly, an arity coercion can be
```

Comment:
I find "parent AST node" a strange concept to introduce here. This is a language design, and not really a compiler design, so it seems a bit circular to talk about AST nodes.

Maybe "moves its parent delimited syntactic construct inside the pack literal"? "deliminited syntactic construct" seems to be a reasonable description to cover the cases you care about here maybe?

Reply:
I've now rephrased with a variant of the suggestion you made offline.